src/Microsoft.AspNetCore.DataProtection/DataProtectionUtilityExtensions.cs

```diff
@@ -36,6 +37,11 @@ public static string GetApplicationUniqueIdentifier(this IServiceProvider servic
             if (services != null)
             {
                 discriminator = services.GetService<IApplicationDiscriminator>()?.Discriminator;
+
+                if (discriminator == null)
+                {
+                    discriminator = services.GetService<IHostingEnvironment>()?.ContentRootPath;
```

Comment:
I liked the new implementation better. Rather than explicitly calling out to IHostingEnvironment, instead register an IApplicationDiscriminator that does it.